### PR TITLE
Initial implementation for inline replies

### PR DESCRIPTION
**Pull request by ikkerens**
Initial implementation for https://github.com/discord/discord-api-docs/pull/2118

I also added in the support for v8, which will need to be uncommented when that gets implemented. Do you guys keep track of a v8 TODO list somewhere?

Full disclaimer though, I have no idea if this works as it is untested vs the Discord API (~~they haven't implemented this yet, so don't merge until they have~~).

### CODE REVIEW

File: src/model/channel/message.rs [hunk 1]
```diff
@@ -879,6 +883,7 @@ enum_number!(
         NitroTier1,
         NitroTier2,
         NitroTier3,
+        // InlineReply, // TODO Enable this when v8 hits
```
Comment (arqunis) [inline src/model/channel/message.rs]: This can be enabled now.

**Review by arqunis — CHANGES_REQUESTED** [review]

**Comment by ikkerens** [comment]
I've enabled the v8 enums, but I can't personally test this inline replies feature with the lib myself atm.

**Review by arqunis — APPROVED** [review]
Besides the [absence of the variant](https://github.com/serenity-rs/serenity/pull/991/files#diff-a95a446fbee2360e72976dd53c83ca8ff33a6319978fe62252ae0bdbb385faf5R906) causing failure in CI, this pull request is fine. I have tested it myself in the Serenity Discord server and works as it should.

**Comment by vicky5124** [comment]
shouldn't Message.reply be updated to use this instead?
i wanna be able to use this extensively  thought out my bot, and having it limited to the message builder doesn't make this convenient.

This is also missing the option to disable the mention to the original author of the message, as possible to be done with the client.

**Comment by Flat** [comment]
I agree reply should be changed to use this, it avoids confusion for new users but may be unexpected behavior for current users. Possibly a notice stating the change prominently in the next release may suffice.

**Comment by Lakelezz** [comment]
My suggestions is to update `reply` to perform the new inline replies, but this is a breaking change as it changes behaviour. Therefore we would have to rebase to the breaking channel.
Additionally, implement `reply_ping`, which sets the toggle for whether the inline reply mentions to `true` and `reply_mention`, which would perform the old school reply we had for a long time (if that's even still needed).

**Comment by kingbri1** [comment]
Agreed with this. These 3 methods give the end user the ability to choose instead of having one firm option.

**Comment by HTG-YT** [comment]
I suggest add another method called `respond` to just send a message normally...?

**Comment by HTG-YT** [comment]
I am not really sure about that 🤣

**Review by Lakelezz — CHANGES_REQUESTED** [review]
As I mentioned above, it would be great to add these please before we can merge it: 
`reply`, performs the inline reply without pinging;
`reply_ping`, performs the inline reply with a ping;
`reply_mention`, performs the original `reply`, such as `@name:`.

**Comment by HTGAzureX1212** [comment]
> As I mentioned above, it would be great to add these please before we can merge it:
> `reply`, performs the inline reply without pinging;
> `reply_ping`, performs the inline reply with a ping;
> `reply_mention`, performs the original `reply`, such as `@name:`.

I don't know if this approach is acceptable: `reply` takes a `ping` `bool` parameter, if `true` is passed in, performs the inline reply with a ping and vice versa.

**Comment by iddm** [comment]
> > As I mentioned above, it would be great to add these please before we can merge it:
> > `reply`, performs the inline reply without pinging;
> > `reply_ping`, performs the inline reply with a ping;
> > `reply_mention`, performs the original `reply`, such as `@name:`.
> 
> I don't know if this approach is acceptable: `reply` takes a `ping` `bool` parameter, if `true` is passed in, performs the inline reply with a ping and vice versa.

This would break the backwards compatibility, and even if we aren't following the semver, this is just not convenient.

**Comment by HTGAzureX1212** [comment]
I meant that to merge `reply` and `reply_ping`, and leave `reply_mention` as is

File: src/model/channel/message.rs [hunk 2]
```diff
@@ -977,6 +983,28 @@ pub struct MessageReference {
     pub(crate) _nonexhaustive: (),
 }
 
+impl From<&Message> for MessageReference {
```
Comment (iddm) [inline src/model/channel/message.rs:986]: Just curious here, isn't `impl From<AsRef<Message>>` better?

**Comment by iddm** [comment]
> I meant that to merge `reply` and `reply_ping`, and leave `reply_mention` as is

I might have been confused a bit, please correct me if I indeed have wrongly understood you.
The [reply method](https://docs.rs/serenity/0.9.1/serenity/model/channel/struct.Message.html#method.reply) has the following interface now in 0.9.1:

```rust
pub async fn reply<'_>(
    &'_ self,
    cache_http: impl CacheHttp,
    content: impl Display
) -> Result<Message>
```
If we are to add a bool parameter there like this
```rust
pub async fn reply<'_>(
    &'_ self,
    ping: bool,
    cache_http: impl CacheHttp,
    content: impl Display
) -> Result<Message>
```
then the current users will have to also add it to all of their invocations of it, what isn't convenient.

**Comment by HTGAzureX1212** [comment]
> > I meant that to merge `reply` and `reply_ping`, and leave `reply_mention` as is
> 
> I might have been confused a bit, please correct me if I indeed have wrongly understood you.
> The [reply method](https://docs.rs/serenity/0.9.1/serenity/model/channel/struct.Message.html#method.reply) has the following interface now in 0.9.1:
> 
> ```rust
> pub async fn reply<'_>(
>     &'_ self,
>     cache_http: impl CacheHttp,
>     content: impl Display
> ) -> Result<Message>
> ```
> 
> If we are to add a bool parameter there like this
> 
> ```rust
> pub async fn reply<'_>(
>     &'_ self,
>     ping: bool,
>     cache_http: impl CacheHttp,
>     content: impl Display
> ) -> Result<Message>
> ```
> 
> then the current users will have to also add it to all of their invocations of it, what isn't convenient.

Was it the original `reply` method in 0.9.1 the new **inline** reply?

What I meant was, the **inline** reply method should have a `ping` `bool` parameter to determine whether to ping the user that was **inline** replied.

**Comment by arqunis** [comment]
To avoid stalling this pull request further, it has been decided to merge it now. The adaptation of the `Message::reply` method will come in a separate commit/pull request.

**Pull request merged**